graph_djikstra_heap.c

```diff
@@ -0,0 +1,240 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#define HEAP_MAX_SIZE 100
+#define HEAP_INIT_SIZE 10
+#define P_QUEUE_ELEMENT_SIZE 2
+
+/*
+pq based on min binary heap, e.g. 10,20,50,30,40,60,70
+heap keeps array of arrays where each element is int* array[2] = [priority, key]
+*/
+
+typedef struct {
+    int **p_queue;
+    int size;
+} Heap;
+
+Heap* heap_create();
+void heap_free(Heap *);
+
+void heap_push(Heap *, int *);
+int* heap_peek(Heap *);
+int* heap_pop(Heap *);
+void heap_print(Heap *);
+
+static void heap_realloc(Heap *);
+static void heap_heapify(Heap *);
+
+void heap_print(Heap* self) {
+    for(int i = 0; i < self->size; i++) {
+        printf("%d ", self->p_queue[i][0]);
+        printf("%d\n", self->p_queue[i][1]);
+    }
+}
+
+int* heap_peek(Heap *self) {
+    return self->p_queue[0];
+}
+
+int* heap_pop(Heap *self) {
+    if (self->size <=0) {
+        perror("Heap is empty");
+        abort();
+    }
+
+    int* popped_arr = self->p_queue[0];
+    self->size--;
+
+    if (self->size <= 0) {
+        return popped_arr;
+    }
+
+    self->p_queue[0] = self->p_queue[self->size];
+
+    int heap_size = self->size;
+    int** heap_values = self->p_queue;
+
+    int index, left, right, smallest;
+    index = 0;
+
+    //As only one node may be move down, the only cycle is sufficient
+    while(1) {
+        heap_print(self);
+        left = 2 * index + 1;
+        right = 2 * index + 2;
+        smallest = index;
+
+        if (left < heap_size && heap_values[left][0] < heap_values[smallest][0]) {
+            smallest = left;
+        }
+
+        if (right < heap_size && heap_values[right][0] < heap_values[smallest][0]) {
+            smallest = right;
+        }
+
+        if (smallest == index) {
+            break;
+        }
+
+        int *tmp = heap_values[index];
+        heap_values[index] = heap_values[smallest];
+        heap_values[smallest] = tmp;
+
+        index = smallest;
+        }
+
+    return popped_arr;
+}
+
+static void heap_realloc(Heap *self) {
+    int new_heap_size = self->size * 2;
+
+    if(new_heap_size > 100) {
+        new_heap_size = HEAP_MAX_SIZE;
+    }
+
+    int **tmp = realloc(self->p_queue, new_heap_size * sizeof(int *));
+
+    if (tmp == NULL) {
+        perror("Reallocation to increase size failed\n");
+        abort();
+    }
+
+    self->p_queue = tmp;
+
+}
+
+void heap_push(Heap *self, int* priority_key_arr) {
+    if(self->size >= HEAP_MAX_SIZE) {
+        perror("Heap is full\n");
+        abort();
+    }
+
+    if((self->size > 0) && (self->size % HEAP_INIT_SIZE == 0)) {
+        heap_realloc(self);
+    }
+
+    self->p_queue[self->size] = priority_key_arr;
+    self->size++;
+
+    if(self->size > 1) {
+        heap_heapify(self);
```

Comment:
looks like from bottom to top traversal should be here instead heapify call?

Reply:
Great thanks!